### PR TITLE
Reorg `Preprocessor Definitions` section.

### PR DESCRIPTION
**Pull request by kito-cheng**
- Tabularization, add category for those macro definitions and describe
   when will define.

### CODE REVIEW

**Comment by ebahapo** [comment]
What exactly is the PIC code model?

File: riscv-c-api.md [hunk 1]
```diff
@@ -33,28 +33,41 @@ https://creativecommons.org/licenses/by/4.0/.
...
-* `__riscv_cmodel_medany`
-* `__riscv_cmodel_pic`
-* `__riscv_vector`
-* `__riscv_bitmanip`
-* `__riscv_zfh`
+| Name                | Value | When defined                  |
+| ------------------- | ----- | ----------------------------- |
+| __riscv             | 1     | Always defined.               |
+| __riscv_xlen        | <ul><li>32 for rv32</li><li>64 for rv64</li><li>128 for rv128</ul> | Always defined.             |
+| __riscv_flen        | <ul><li>128 if `Q` extension available</li><li>64 if `D` extension available</li><li>32 if `F` extension available<li></ul> | `F` extension is available. |
```
Comment (ebahapo) [inline riscv-c-api.md:40]: Or rather:

- 32 if the `F` extension is available **or**
- 64 if the `D` extension is available **or**
- 128 if the `Q` extension is available

**Comment by kito-cheng** [comment]
> What exactly is the PIC code model?

In fact we treat PIC/PIE as a code model like `medlow` and `medany` in the GCC implementation, which mean if you apply -fPIC/-fpic/-fPIE/-fpie then -mcmodel is no longer matters.

This also remind me that clang don't have same behavior on this macro.

File: riscv-c-api.md (hunk 1 above)
Comment (kito-cheng) [reply riscv-c-api.md]: Updated, thanks :)

**Comment by jim-wilson** [comment]
The pic code model is medany with different symbol resolution, to make shared libraries work.  It is a bit of a gcc hack to call it a code model.  That perhaps made the implementation a little simpler and/or made the compiler a little easier to use.  I always considered this an internal gcc implementation detail.  I don't think that we should be exposing this outside gcc.  Though I see that we already have riscv_cmodel_pic.  Maybe we could deprecate that if no one is using it.  If you want to know if code was compiled pic, you should be using __PIC__ or __pic__ or __PIE__ or __pie__.

**Comment by lenary** [comment]
I think we could just leave `__riscv_cmodel_pic` undocumented here, and potentially ensure that when in PIC-mode, gcc defines `__riscv_cmodel_medany`?

It might be good to add additional notes to `_div`, `_fdiv`, `_fsqrt` which link them to the related GCC flags. Clang unconditionally defines them all when the respective extension is enabled, because we don't have the `-mno-div` flags and the like.

**Comment by ebahapo** [comment]
Ok, it seems to me then that it would be better to not add the predefined macro `__riscv_cmodel_pic` here then.  And I agree with @lenary , that PIC should then define `__riscv_cmodel_medany` as well.

Besides, PIC already defines `__PIC__`, as does PIE, which defines both the latter and `__PIE__`.

**Comment by kito-cheng** [comment]
ChangeLog
- Move `__riscv_cmodel_pic` to `Deprecated Preprocessor Definitions`.
- Mention `__riscv_div`, `__riscv_muldiv`, `__riscv_fdiv` and `__riscv_fsqrt` won't defined when `-mno-div` and `-mno-fdiv` option is given

File: riscv-c-api.md [hunk 2]
```diff
@@ -33,28 +33,47 @@ https://creativecommons.org/licenses/by/4.0/.
...
+| Name                  | Value    | When defined                          |
+| --------------------- | -------- | ------------------------------------- |
+| __riscv_cmodel_medlow | 1        | Defined if using `medlow` code model. |
+| __riscv_cmodel_medany | 1        | Defined if using `medany` code model. |
+
+### Deprecated Preprocessor Definitions
+
+| Name                  | Value    | When defined                          | Alternative |
+| --------------------- | -------- | ------------------------------------- | ----------- |
+| __riscv_cmodel_pic    | 1        | GCC will defined that when program compile with `-fPIC`, `-fpic`, `-fPIE` or `-fpie`. | `__PIC__` or `__PIE__` |
```
Comment (lenary) [inline riscv-c-api.md:76]:
```suggestion
| __riscv_cmodel_pic    | 1        | GCC defines this when compiling with `-fPIC`, `-fpic`, `-fPIE` or `-fpie`. | `__PIC__` or `__PIE__` |
```

**Review by lenary — APPROVED** [review]
One grammar nit, then LGTM.

File: riscv-c-api.md (hunk 2 above)
Comment (kito-cheng) [reply riscv-c-api.md]: Cool! this is the first time I have seen this github feature :P

**Comment by kito-cheng** [comment]
@jim-wilson ping

File: riscv-c-api.md [hunk 3]
```diff
@@ -33,28 +33,47 @@ https://creativecommons.org/licenses/by/4.0/.
...
+| __riscv_mul         | 1     | `M` extension is available.   |
+| __riscv_div         | 1     | `M` extension is available and `-mno-div` is not given.*[1]    |
+| __riscv_muldiv      | 1     | `M` extension is available and `-mno-div` is not given.*[1]    |
+| __riscv_fdiv        | 1     | `F` extension is available and `-mno-fdiv` is not given.*[1]   |
+| __riscv_fsqrt       | 1     | `F` extension is available and `-mno-fdiv` is not given.*[1]   |
+| __riscv_vector      | 1     | `V` extension is available.   |
+| __riscv_bitmanip    | 1     | `B` extension is available.   |
+| __riscv_zfh         | 1     | `Zfh` extension is available. |
+
+*[1] Not all compiler provide `-mno-div` and `-mno-fdiv` option.
```
Comment (jim-wilson) [inline riscv-c-api.md:53]:
```suggestion
*[1] Not all compilers provide `-mno-div` and `-mno-fdiv` option.
```

**Review by jim-wilson — APPROVED** [review]
Looks good with one minor English improvement.

This will require a gcc patch to define __riscv_cmodel_medany when PIC.  Are you writing that patch or am I?

**Comment by kito-cheng** [comment]
@jim-wilson 
> This will require a gcc patch to define __riscv_cmodel_medany when PIC. Are you writing that patch or am I?
I am writing that, I'll send that today :)

**Comment by vineetgarc** [comment]
> > What exactly is the PIC code model?
> 
> In fact we treat PIC/PIE as a code model like `medlow` and `medany` in the GCC implementation, which mean if you apply -fPIC/-fpic/-fPIE/-fpie then -mcmodel is no longer matters.

@kito-cheng Is this still true ?

**Comment by kito-cheng** [comment]
>> > What exactly is the PIC code model?
>>
>> In fact we treat PIC/PIE as a code model like medlow and medany in the GCC implementation, which mean if you apply -fPIC/-fpic/-fPIE/-fpie then -mcmodel is no longer matters.
>
>@kito-cheng Is this still true ?

We didn't change those part for a long time, so, yes.

**Pull request merged**